adds a simple encoder model

RussTedrake opened the pull request:
with quantization and calibration offsets.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/4247)
<!-- Reviewable:end -->

RussTedrake commented:
+@jwnimmer-tri for platform review, +@hongkai-dai for feature review

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXHfRoCp35HPP9WRw_E:bxhtojv)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 9 of 10 files at r1.
Review status: 7 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/autodiff_overloads.h, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHhzH32MEudjektOEm:-KXHhzH32MEudjektOEn:b50hi7p) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L64)):*
> ```objc
> Eigen::AutoDiffScalar<DerTypeA> copysign(
>     const Eigen::AutoDiffScalar<DerTypeA>& x, const DerTypeB& y) {
>   if (x * y < 0)
> ```

What if we have an overflow or underflow problem by doing multiplication? Can we check the sign of `x` and `y` separately, and return `x` of  `-x` depending on whether they have the same sign?

---

*[drake/systems/sensors/encoders.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHviyPGE9WB-IipsJt:-KXHviyPGE9WB-IipsJu:boexxga) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L16)):*
> ```objc
> /// to capture the sampled-data effects.
> ///
> /// @ingroup sensor_systems
> ```

Add a documentation on the quantization and offset? If I read correctly, the quantization we use is different from Mathworks https://www.mathworks.com/help/comm/ref/quantizingencoder.html, which is more like
```
ceiling((y(i)/M_2_PI)*ticks_per_revolution[i]) - 1
```
We might want to explain why we choose a different quantization from Mathworks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
+(status: curate commits before merging)

Checkpoint of review on the preparatory work.  On to the encoder itself next.

---

Reviewed 7 of 10 files at r1.
Review status: 7 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/common/autodiff_overloads.h, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHv17J2rXsad1MsRpG:-KXHv17J2rXsad1MsRpH:b4mtg4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L5)):*
> ```objc
> ///
> /// Used via argument-dependent lookup (ADL). These functions appear
> /// in the global namespace so that ADL can automatically choose between
> ```

The comment about "global namespace" and related explanation here is incorrect now.

---

*[drake/common/autodiff_overloads.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHvqSEeadH9eSVLXGw:-KXHvqSFLMKKGjIEFE-D:b-ufk8ub) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L61)):*
> ```objc
> 
> /// Overloads copysign from <cmath>.
> template <typename DerTypeA, typename DerTypeB>
> ```

BTW `DerTypeB` is slightly misleading; I think `DerType` usually refers to the thing within the `AutoDiffScalar`, not a full `ScalarType` or `double` as this allows for.  (I don't have a better suggestion, though.)

---

*[drake/common/autodiff_overloads.h, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHw8ixzHS1n4ZcKIqb:-KXHw8ixzHS1n4ZcKIqc:b-wx6mxt) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L68)):*
> ```objc
>   else
>     return x;
> }
> ```

This implementation (and the one below) do not appear to handle NaNs the same way as `copysign` in `<cmath>` does.  (A `copysign(nan, -1)` should be `-nan` but this implementation produces `nan`; `copysign(1, -nan`) should be `-1` but this implementation produces `1`.)

That should either be fixed, or (perhaps more sensibly) replaced with a fail-fast assertion (that neither x nor y are NaN) and/or noted with an API comment.

---

*[drake/systems/analysis/sensors.h, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHuY0b8n8-TUHs2Lng:-KXHuY0dAgTTdh7r5ABw:byu87ks) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/analysis/sensors.h#L6)):*
> ```objc
>  * @ingroup systems
>  * @}
>  */
> ```

BTW I guess this is fine, but why place the file in this directory?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXHuvJSxRO35OOvWmzx:b-id9eiz)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Checkpoint of the encoder.  On to the test next.

---

Reviewed 1 of 10 files at r1, 2 of 2 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/encoders.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI3uBanbjnXPv8Brgr:-KXI3uBanbjnXPv8Brgs:b-muqnwy) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L61)):*
> ```cpp
>                                    systems::SystemOutput<T>* output) const {
>   auto y = output->GetMutableVectorData(0)->get_mutable_value();
>   const auto& calibration_offset =
> ```

BTW I think this should be `calibration_offsets` (plural)?  I was initially confused how they all shared an offset (the `auto` hid the fact that it was vector-valued instead of scalar-valued).

---

*[drake/systems/sensors/encoders.cc, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI4SQS_u4nb3LMaoBd:-KXI4SQS_u4nb3LMaoBe:b-ivgn77) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L62)):*
> ```cpp
>   auto y = output->GetMutableVectorData(0)->get_mutable_value();
>   const auto& calibration_offset =
>       this->GetNumericParameter(context, 0).get_value();
> ```

BTW Consider asserting that the dimension of the offsets vector is correct (num_encoders).

---

*[drake/systems/sensors/encoders.cc, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI4owHVJ3AEtb7KSl0:-KXI4owJWt2ki00llpt5:b-x8ff0k) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L69)):*
> ```cpp
> 
>     // Calibration.
>     y(i) = this->EvalVectorInput(context, 0)->GetAtIndex(index) -
> ```

BTW `EvalVectorInput` is _supposed_ to be very cheap, but consider manually hoisting it out of the loop, near the y and offsets?

---

*[drake/systems/sensors/encoders.cc, line 77 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI5F_OOr85yiPwigfd:-KXI5F_OOr85yiPwigfe:bpsik0k) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L77)):*
> ```cpp
>       using std::floor;
>       using std::copysign;
>       // Round towards zero
> ```

BTW nit: Missing period.

---

*[drake/systems/sensors/encoders.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI6y23JWuzj0_SekfB:-KXI6y24RU0s2jJQuZvI:b-g7z98r) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L91)):*
> ```cpp
>   return std::make_unique<systems::Parameters<T>>(
>       std::make_unique<systems::BasicVector<T>>(
>           Eigen::Matrix<T, Eigen::Dynamic, 1>::Zero(num_encoders_)));
> ```

BTW Consider `VectorX<T>::Zero(num_encoders_)` to replace `Eigen::Matrix<T, Eigen::Dynamic, 1>::Zero(num_encoders_)` for brevity?

---

*[drake/systems/sensors/encoders.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHyReGZTmplklq7H-I:-KXHyReHLvbhiE-W8Xtu:b-9ws7ie) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L23)):*
> ```objc
>   /// port.
>   explicit RotaryEncoders(
>       const std::vector<unsigned int>& ticks_per_revolution);
> ```

META Can't use `unsigned` as a type anywhere in this class, per https://google.github.io/styleguide/cppguide.html#Integer_Types

(I suspect a plain `int` is the better answer.)

---

*[drake/systems/sensors/encoders.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHzHCD_0MSpdnVccRx:-KXHzHCD_0MSpdnVccRy:b-f1jh3t) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L28)):*
> ```objc
>   /// use in the output.
>   /// @param input_port_size Dimension of the expected input signal
>   /// @param input_vector_indices List of indices
> ```

From this API doc, it's not clear to me what `ticks_per_revolution` the provided inputs will be encoded using.  I assumed "1" but in fact its "none".

---

*[drake/systems/sensors/encoders.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHyq6Ga4YUqQxM9B0_:-KXHyq6Ga4YUqQxM9B0a:b-imvwaz) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L29)):*
> ```objc
>   /// @param input_port_size Dimension of the expected input signal
>   /// @param input_vector_indices List of indices
>   RotaryEncoders(const int input_port_size,
> ```

Per code_style_guide, in API declarations with pass-by-value (like `int input_port_size` here), don't use the meaningless `const` because it makes misleading noise in the doxygen.  (The definition in the cc file can keep the const.)

Same for the below constuctor.

---

*[drake/systems/sensors/encoders.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI0I97E11oLTkuSIs7:-KXI0I97E11oLTkuSIs8:bapkrev) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L36)):*
> ```objc
>                  const std::vector<unsigned int>& input_vector_indices,
>                  const std::vector<unsigned int>& ticks_per_revolution);
> ```

Per cppguide, copy and assignment must be explicitly deleted (in public scope), for example https://github.com/RobotLocomotion/drake/blob/f86b8348c9b060579fea971db22d739f9ece43a9/drake/systems/framework/primitives/multiplexer.h#L38-L40

---

*[drake/systems/sensors/encoders.h, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI12dkwn300xXK6hNw:-KXI12dlfyzwjyLci1fi:b-5vbn0a) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L54)):*
> ```objc
> 
>  private:
>   const unsigned int num_encoders_;          /// Dimension of the output port.
> ```

Missing initializer (e.g., `{}` between `_` and `;`) per code_style_guide.

---

*[drake/systems/sensors/encoders.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI-REzafLOh0FWSaCb:-KXI-RF-U3lw0jmjR7p3:bjzs8cu) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L61)):*
> ```objc
> }  // namespace sensors
> }  // namespace systems
> }  // namespace drake
> ```

I am a little skeptical of calling this `encoders.h`, as if more classes were going to be added here?  It seems like `rotary_encoders.h` would be more typical?  (One class per file should be the default, unless the two classes truly can't be separated.)

https://google.github.io/styleguide/cppguide.html#File_Names is the guide's advice.

---

*[drake/systems/sensors/encoders.h, line 14 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI9iHA7CyGOJnGbKtf:-KXI9iHA7CyGOJnGbKtg:b-4cju3f) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.h#L14)):*
> ```objc
> /// Simple model to capture the quantization and calibration offset effects
> /// of a rotary encoder.  Consider combining this with a zero-order hold system
> /// to capture the sampled-data effects.
> ```

BTW Until I read the code, it wasn't clear to me that the output port units were still radians.  (I assumed it was in ticks.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXIAK71lPezQvvxjnlK:b5088pj)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
First pass done.

---

Reviewed 1 of 10 files at r1.
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/test/encoders_test.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIB6OsgtmE06J0Jlz8:-KXIB6OsgtmE06J0Jlz9:b-xfbhwj) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L10)):*
> ```cpp
> #include "drake/common/eigen_matrix_compare.h"
> 
> namespace drake {
> ```

(Within drake) please enclose all test code within an anonymous namespace.

---

*[drake/systems/sensors/test/encoders_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIAbS6_6Yc5f9LMizh:-KXIAbS6_6Yc5f9LMizi:b8jarnz) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L12)):*
> ```cpp
> namespace drake {
> 
> // Test with the simple quantization-only constructor.
> ```

BTW "quantization-only constructor" and "selector-only constructor" are nicely descriptive.  Perhaps the ctor APIs should use those same terms?

---

*[drake/systems/sensors/test/encoders_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXICzT-wLD6Obqfw_9e:-KXICzT0IRU9CKnx-gQS:b-o0h5rv) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L29)):*
> ```cpp
>   Eigen::Vector2d angle, desired_measurement;
> 
>   srand(42);
> ```

We have a strong preference for repeatable tests:
http://drake.mit.edu/code_review_checklist.html#is-your-code-deterministic

Eigen can use C++11 random generators:
https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a997ef3960458fb5dedb8f7f78b068a9b

Since we don't have great tooling or examples of this stood up, I could probably stand a TODO & issue here, to repair it later.  Or you can be the first one to do test-randomness correctly!

(I have some small recollection that googletest can also make randomness-injection more structured.)

---

*[drake/systems/sensors/test/encoders_test.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIBxA_BBnk5La4HTIw:-KXIBxA_BBnk5La4HTIx:bdftb7w) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L30)):*
> ```cpp
> 
>   srand(42);
>   for (unsigned int i = 0; i < 10; i++) {
> ```

META No `unsigned`s allowed here, either.

---

*[drake/systems/sensors/test/encoders_test.cc, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEgxofENEMzSo2Udo:-KXIEgxp6bw5J8UMStte:b-7gfty) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L37)):*
> ```cpp
>       if (angle(j) < 0.0)
>         desired_measurement(j) =
>             ceil(angle(j) * ticks_per_radian(j)) / ticks_per_radian(j);
> ```

META Multi-line if-else bodies all require curly braces.

---

*[drake/systems/sensors/test/encoders_test.cc, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIDsfNOxqHL804qydf:-KXIDsfOAyKoqDyKvmpG:b-dfi062) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L55)):*
> ```cpp
> GTEST_TEST(TestEncoders, SelectorOnly) {
>   // Construct a system with no quantization, and only inputs 2 and 3 are
>   // passed.
> ```

BTW Only inputs _1_ and _2_ are passed, right?

---

*[drake/systems/sensors/test/encoders_test.cc, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEOgo9XccO8iJK9LZ:-KXIEOgo9XccO8iJK9L_:b-qcufrt) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L95)):*
> ```cpp
>   auto measurement = output->get_vector_data(0);
> 
>   // set some offsets
> ```

cappunc

---

*[drake/systems/sensors/test/encoders_test.cc, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEGKqFx73SYVmiRpQ:-KXIEGKqFx73SYVmiRpR:bcnz74p) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L128)):*
> ```cpp
> }
> 
> }  // namespace drake
> ```

Missing unit test coverage of the 3-arg encoder constructor, I think?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXIG7g-vOq6eNYOPv8x:b-te4kzy)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: all files reviewed at latest revision, 26 unresolved discussions, some commit checks failed.

---

*[drake/common/autodiff_overloads.h, line 5 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHv17J2rXsad1MsRpG:-KXIDnAJHC5cRUJwVtcO:b-cypxrc) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L5)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The comment about "global namespace" and related explanation here is incorrect now.
</details>
updated.

---

*[drake/common/autodiff_overloads.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHvqSEeadH9eSVLXGw:-KXIE12J95mksQC-xvLK:b-qqzcd4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW `DerTypeB` is slightly misleading; I think `DerType` usually refers to the thing within the `AutoDiffScalar`, not a full `ScalarType` or `double` as this allows for.  (I don't have a better suggestion, though.)
</details>
replaced it with DerType and T.

---

*[drake/common/autodiff_overloads.h, line 64 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHhzH32MEudjektOEm:-KXIFl8ThAZEIF47fUHV:b-cypxrc) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L64)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> What if we have an overflow or underflow problem by doing multiplication? Can we check the sign of `x` and `y` separately, and return `x` of  `-x` depending on whether they have the same sign?
</details>
updated.

---

*[drake/common/autodiff_overloads.h, line 68 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHw8ixzHS1n4ZcKIqb:-KXIFm8KGoSosie9ttJu:b-cypxrc) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/common/autodiff_overloads.h#L68)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> This implementation (and the one below) do not appear to handle NaNs the same way as `copysign` in `<cmath>` does.  (A `copysign(nan, -1)` should be `-nan` but this implementation produces `nan`; `copysign(1, -nan`) should be `-1` but this implementation produces `1`.)
> 
> That should either be fixed, or (perhaps more sensibly) replaced with a fail-fast assertion (that neither x nor y are NaN) and/or noted with an API comment.
</details>
updated.

---

*[drake/systems/analysis/sensors.h, line 6 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHuY0b8n8-TUHs2Lng:-KXIFoWt5X0qT75HnDaQ:b-c1dzox) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/analysis/sensors.h#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I guess this is fine, but why place the file in this directory?
</details>
bizarre.  must have been user-error in my IDE. good catch.  relocated to systems/sensors/

---

*[drake/systems/sensors/encoders.cc, line 61 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI3uBanbjnXPv8Brgr:-KXIGb2nUIFd9IcrNWu5:b-rmo4oj) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I think this should be `calibration_offsets` (plural)?  I was initially confused how they all shared an offset (the `auto` hid the fact that it was vector-valued instead of scalar-valued).
</details>
done.  and replaced the auto with the type.

---

*[drake/systems/sensors/encoders.cc, line 62 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI4SQS_u4nb3LMaoBd:-KXIGhPrS4k_V8zSZvEw:bydt6g6) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L62)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider asserting that the dimension of the offsets vector is correct (num_encoders).
</details>
I didn't think it was needed because I have the setter method (below).  but i suppose someone could go around that.  have added it.

---

*[drake/systems/sensors/encoders.cc, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI4owHVJ3AEtb7KSl0:-KXIHJ1XlAFxRhwATAcO:b-7rl0jd) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW `EvalVectorInput` is _supposed_ to be very cheap, but consider manually hoisting it out of the loop, near the y and offsets?
</details>
done.

---

*[drake/systems/sensors/encoders.cc, line 77 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI5F_OOr85yiPwigfd:-KXIHLSPc1rMz0xCH-9I:bb7f2mf) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L77)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW nit: Missing period.
</details>
done

---

*[drake/systems/sensors/encoders.cc, line 91 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI6y23JWuzj0_SekfB:-KXIHYSia6p4iOGkr0zw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.cc#L91)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider `VectorX<T>::Zero(num_encoders_)` to replace `Eigen::Matrix<T, Eigen::Dynamic, 1>::Zero(num_encoders_)` for brevity?
</details>
Done.

---

*[drake/systems/sensors/encoders.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHviyPGE9WB-IipsJt:-KXIIHif5EnT_ziyumJj:b-xc7ida) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L16)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary>
> Add a documentation on the quantization and offset? If I read correctly, the quantization we use is different from Mathworks https://www.mathworks.com/help/comm/ref/quantizingencoder.html, which is more like
> ```
> ceiling((y(i)/M_2_PI)*ticks_per_revolution[i]) - 1
> ```
> We might want to explain why we choose a different quantization from Mathworks.
</details>
interesting.  that is what I implemented at first, but i think the model i've used here is more correct... no?  (if you go negative, you don't get any ticks until you've crossed the first one). i've updated the doc.

---

*[drake/systems/sensors/encoders.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHyReGZTmplklq7H-I:-KXIIxR8_xV2wsDqBVEq:b-jbg1kn) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META Can't use `unsigned` as a type anywhere in this class, per https://google.github.io/styleguide/cppguide.html#Integer_Types
> 
> (I suspect a plain `int` is the better answer.)
</details>
did not know that.  updated, and added a bunch more DRAKE_ASSERTs to check the positivity of my input arguments.

---

*[drake/systems/sensors/encoders.h, line 28 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHzHCD_0MSpdnVccRx:-KXIKY1bx-RPsyaa6EiU:b-cypxrc) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L28)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> From this API doc, it's not clear to me what `ticks_per_revolution` the provided inputs will be encoded using.  I assumed "1" but in fact its "none".
</details>
updated.

---

*[drake/systems/sensors/encoders.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHyq6Ga4YUqQxM9B0_:-KXIKe6f4UVVLs6s_H8Z:b-nks453) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Per code_style_guide, in API declarations with pass-by-value (like `int input_port_size` here), don't use the meaningless `const` because it makes misleading noise in the doxygen.  (The definition in the cc file can keep the const.)
> 
> Same for the below constuctor.
</details>
that one I knew.  my bad.

---

*[drake/systems/sensors/encoders.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI0I97E11oLTkuSIs7:-KXIKw_p8qOTj75ZH5vg:beanskd) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L36)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Per cppguide, copy and assignment must be explicitly deleted (in public scope), for example https://github.com/RobotLocomotion/drake/blob/f86b8348c9b060579fea971db22d739f9ece43a9/drake/systems/framework/primitives/multiplexer.h#L38-L40
</details>
I've zapped them.  But I just read the style guide.  Why must they be deleted?  There is nothing locally that would appear to go wrong with the default copy and assigment constructors?

---

*[drake/systems/sensors/encoders.h, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI12dkwn300xXK6hNw:-KXIM-EdhS4lUEF69IuW:b-bjo4of) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing initializer (e.g., `{}` between `_` and `;`) per code_style_guide.
</details>
Done.  but just to make sure I understand... why?  Icouldn't see it in the style-guide, and it is always initialized via a constructor.  

---

*[drake/systems/sensors/encoders.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI-REzafLOh0FWSaCb:-KXIMMzKkG4AT-WVMomO:b-t8sdc8) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/encoders.h#L61)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> I am a little skeptical of calling this `encoders.h`, as if more classes were going to be added here?  It seems like `rotary_encoders.h` would be more typical?  (One class per file should be the default, unless the two classes truly can't be separated.)
> 
> https://google.github.io/styleguide/cppguide.html#File_Names is the guide's advice.
</details>
done.  i was planning to have linear encoders in here eventually, and would have preferred less compilation units.  is one class per file really the best practice?

---

*[drake/systems/sensors/encoders.h, line 14 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI9iHA7CyGOJnGbKtf:-KXIHw4YDTL3xQ4LDBAT:bk3ag8w) ([raw file](https://github.com/robotlocomotion/drake/blob/29ca3e68621338e2f020d7881ab4d5c2d8acc9db/drake/systems/sensors/encoders.h#L14)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Until I read the code, it wasn't clear to me that the output port units were still radians.  (I assumed it was in ticks.)
</details>
updated doc.

---

*[drake/systems/sensors/test/encoders_test.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIB6OsgtmE06J0Jlz8:-KXINR5PTRYgWG0tNWXC:bhtng3) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> (Within drake) please enclose all test code within an anonymous namespace.
</details>
done.  but i thought most tests were put in the drake namespace? (or even something more specific)?

---

*[drake/systems/sensors/test/encoders_test.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIAbS6_6Yc5f9LMizh:-KXIN4A_SND2W7sW66R-:b-7rl0jd) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW "quantization-only constructor" and "selector-only constructor" are nicely descriptive.  Perhaps the ctor APIs should use those same terms?
</details>
done.

---

*[drake/systems/sensors/test/encoders_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXICzT-wLD6Obqfw_9e:-KXINuEK2ozj26ZLliUb:b-ufnvkx) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> We have a strong preference for repeatable tests:
> http://drake.mit.edu/code_review_checklist.html#is-your-code-deterministic
> 
> Eigen can use C++11 random generators:
> https://eigen.tuxfamily.org/dox/classEigen_1_1DenseBase.html#a997ef3960458fb5dedb8f7f78b068a9b
> 
> Since we don't have great tooling or examples of this stood up, I could probably stand a TODO & issue here, to repair it later.  Or you can be the first one to do test-randomness correctly!
> 
> (I have some small recollection that googletest can also make randomness-injection more structured.)
</details>
i explicitly seed the random number generator above.  why do we say it's ok to use and seed a c++11 random generator, but not the Eigen (which uses rand) generators?  I've written code with the c++11 random generators before, but it's syntactically very painful.

---

*[drake/systems/sensors/test/encoders_test.cc, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIBxA_BBnk5La4HTIw:-KXINhVRz-u1Sctsv1b-:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L30)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META No `unsigned`s allowed here, either.
</details>
Done.

---

*[drake/systems/sensors/test/encoders_test.cc, line 37 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEgxofENEMzSo2Udo:-KXIPHc8-vWkpAtCRwcx:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L37)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> META Multi-line if-else bodies all require curly braces.
</details>
Done.

---

*[drake/systems/sensors/test/encoders_test.cc, line 55 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIDsfNOxqHL804qydf:-KXIOCOEo2fY15-FnyZi:b-j3z6ya) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L55)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Only inputs _1_ and _2_ are passed, right?
</details>
depending how you count.  :)

---

*[drake/systems/sensors/test/encoders_test.cc, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEOgo9XccO8iJK9LZ:-KXIP2kH_RDDNcmu1kf1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L95)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> cappunc
</details>
Done.

---

*[drake/systems/sensors/test/encoders_test.cc, line 128 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEGKqFx73SYVmiRpQ:-KXIP2C06KQp7dRIkbW0:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/encoders_test.cc#L128)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Missing unit test coverage of the 3-arg encoder constructor, I think?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 5 of 11 files at r3.
Review status: 7 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[drake/systems/sensors/rotary_encoders.h, line 36 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI0I97E11oLTkuSIs7:-KXIS56otjpMrjX2uyd8:b1thorl) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/rotary_encoders.h#L36)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> I've zapped them.  But I just read the style guide.  Why must they be deleted?  There is nothing locally that would appear to go wrong with the default copy and assigment constructors?
</details>
Ah, I wasn't 100% clear.  The thing to avoid is being ambiguous or inconsistent.  Inconsistent would be defining copy but not assign for example, but that's not a common mistake.  Ambugious is the user not knowing if copy (and assign) are supported.  So if you want copyable that's fine, but you should write that down if so, either with `= default;` (instead of `= delete;`) or API comment(s), or both.  And also test that copy/assign works okay (which is why we usually `= delete` unless there's a specific need).  Also, I think `LeafSystem` isn't copyable, so you probably can't copy here anyway.

---

*[drake/systems/sensors/rotary_encoders.h, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI12dkwn300xXK6hNw:-KXITOmyWmW85KZTkwBt:b-x1zne2) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/rotary_encoders.h#L54)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Done.  but just to make sure I understand... why?  Icouldn't see it in the style-guide, and it is always initialized via a constructor.  
</details>
http://drake.mit.edu/code_style_guide.html#additional-rules second bullet
Basically "its easy to do by rote" and "having garbage is bad" and "less multi-file reasoning required".

---

*[drake/systems/sensors/rotary_encoders.h, line 61 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI-REzafLOh0FWSaCb:-KXITnSMPozAtJ5NEklI:b-jyncy5) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/rotary_encoders.h#L61)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> done.  i was planning to have linear encoders in here eventually, and would have preferred less compilation units.  is one class per file really the best practice?
</details>
I would say so, yes.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIB6OsgtmE06J0Jlz8:-KXIUA80-WZl58IuCwuT:b-fh1mdb) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L10)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> done.  but i thought most tests were put in the drake namespace? (or even something more specific)?
</details>
Not done (missing push)?

The idea is this:
```
namespace drake {
namespace {
```

That way, the test code gets private linkage, but you still can avoid saying `drake::` everywhere.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXICzT-wLD6Obqfw_9e:-KXIUmNqgiQvClB3c8aj:b-mkwmy7) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L29)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> i explicitly seed the random number generator above.  why do we say it's ok to use and seed a c++11 random generator, but not the Eigen (which uses rand) generators?  I've written code with the c++11 random generators before, but it's syntactically very painful.
</details>
We can decide to agree with this compromise, but (1) Eigen doesn't say which rand it uses, so might not always use `rand()`; (2) in non-test code at least, it's harder to reason about what code buried somewhere called `rand()` and thus upset our sequence to be non-deterministic.  I would also be happier if we had a random sequence that was identical on all platforms, like one of the mt's, so that test failures showed up always (or never), without me having to go hunt down whatever the rand() did on OSX or something.  But better printing of, e.g., the input`angle` here could mitigate that.

In any case, I wouldn't imagine directly calling std random stuff, I agree that's verbose -- I imagine having a test factory on the stack that builds randomized vectors from a fixed seed.

```
"#include drake/common/test/random_matrix_generator.h"
...
{
  test::random_matrix_generator gen;
  ...
  for (...) {
    angle = gen.Make(2);
    mat = gen.Make(2, 4);
```
or even a gtest contraption that builds that in, including the "do this 10 times" (or whatever) boilerplate.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEOgo9XccO8iJK9LZ:-KXIU7LQI7mwpmWYZ8Sp:b-q6df16) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L95)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> Done.
</details>
Not done (missing push)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 11 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 176 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIXLuZcQ2Y6FEyUuDq:-KXIXLuZcQ2Y6FEyUuDr:b-2e6y7a) ([raw file](https://github.com/robotlocomotion/drake/blob/777499143b15bb3f0c2b29ad0548bf9af6b1eca5/drake/systems/sensors/test/rotary_encoders_test.cc#L176)):*
> ```cpp
> }
> 
> }  // namespace drake
> ```

BTW To my read, the unit test file is pushing the border of being overly duplicative.  Consider whether factoring out common code would yield net-positive effort in maintaining this test for the future.  (e.g., A method that given an Encoder + angle + desired_measurement, runs the case using the System 2.0 verbiage; runs an experiment 10 times given a lambda from angle-to-desired, etc.)  I could see leaving it alone being the best answer, or a really beautiful compression.  Hard to say without trying it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXI_9ApPemlt3R7BpmX:bnfp4nl)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 9 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/systems/sensors/rotary_encoders.h, line 54 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXI12dkwn300xXK6hNw:-KXIgIOjWEQsqfZ1mMjY:b-v20yo4) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/rotary_encoders.h#L54)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> http://drake.mit.edu/code_style_guide.html#additional-rules second bullet
> Basically "its easy to do by rote" and "having garbage is bad" and "less multi-file reasoning required".
</details>
got it.  was looking in the google style guide.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 10 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIB6OsgtmE06J0Jlz8:-KXIgLNRxNj5RRLOW-G6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L10)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Not done (missing push)?
> 
> The idea is this:
> ```
> namespace drake {
> namespace {
> ```
> 
> That way, the test code gets private linkage, but you still can avoid saying `drake::` everywhere.
</details>
Done.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXICzT-wLD6Obqfw_9e:-KXIgNBv3BzmrG32N169:b-g9mmzm) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L29)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> We can decide to agree with this compromise, but (1) Eigen doesn't say which rand it uses, so might not always use `rand()`; (2) in non-test code at least, it's harder to reason about what code buried somewhere called `rand()` and thus upset our sequence to be non-deterministic.  I would also be happier if we had a random sequence that was identical on all platforms, like one of the mt's, so that test failures showed up always (or never), without me having to go hunt down whatever the rand() did on OSX or something.  But better printing of, e.g., the input`angle` here could mitigate that.
> 
> In any case, I wouldn't imagine directly calling std random stuff, I agree that's verbose -- I imagine having a test factory on the stack that builds randomized vectors from a fixed seed.
> 
> ```
> "#include drake/common/test/random_matrix_generator.h"
> ...
> {
>   test::random_matrix_generator gen;
>   ...
>   for (...) {
>     angle = gen.Make(2);
>     mat = gen.Make(2, 4);
> ```
> or even a gtest contraption that builds that in, including the "do this 10 times" (or whatever) boilerplate.
</details>
fwiw. not official documentation, but:
https://listengine.tuxfamily.org/lists.tuxfamily.org/eigen/2010/07/msg00403.html

a factory could be good for the wish list.  compromise: will add a todo.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 95 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIEOgo9XccO8iJK9LZ:-KXIg1ehtuHPF5YgDaS6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/test/rotary_encoders_test.cc#L95)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Not done (missing push)?
</details>
Done.

---

*[drake/systems/sensors/test/rotary_encoders_test.cc, line 176 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXIXLuZcQ2Y6FEyUuDq:-KXIg3_PYxLlPRcNjZk4:b2dzcr6) ([raw file](https://github.com/robotlocomotion/drake/blob/777499143b15bb3f0c2b29ad0548bf9af6b1eca5/drake/systems/sensors/test/rotary_encoders_test.cc#L176)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW To my read, the unit test file is pushing the border of being overly duplicative.  Consider whether factoring out common code would yield net-positive effort in maintaining this test for the future.  (e.g., A method that given an Encoder + angle + desired_measurement, runs the case using the System 2.0 verbiage; runs an experiment 10 times given a lambda from angle-to-desired, etc.)  I could see leaving it alone being the best answer, or a really beautiful compression.  Hard to say without trying it.
</details>
agreed.  but i'll favor the (lazier) leave-it-alone solution for now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
@hongkai-dai -- hoping for your ok

---

Review status: 9 of 10 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXIhh5476eXxazKBQo-:bgj3zky)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 11 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXIiJJtSCu8kngBoDBX:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 3 of 11 files at r3, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/systems/sensors/rotary_encoders.cc, line 81 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXJ2XPJ9BGj0mBSOpfi:-KXJ2XPK6TW7Pg99491Z:b71oesd) ([raw file](https://github.com/robotlocomotion/drake/blob/a11555cfaaafe56648b8fdba3edf5753eb571466/drake/systems/sensors/rotary_encoders.cc#L81)):*
> ```cpp
>     // Quantization.
>     if (!ticks_per_revolution_.empty()) {
>       using std::abs;
> ```

Unused now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/sensors/rotary_encoders.cc, line 81 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXJ2XPJ9BGj0mBSOpfi:-KXJ5aFsxqs-LKvMeEhM:b-330k1n) ([raw file](https://github.com/robotlocomotion/drake/blob/a11555cfaaafe56648b8fdba3edf5753eb571466/drake/systems/sensors/rotary_encoders.cc#L81)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Unused now.
</details>
Done.  (my autodiff overload is also unused now, except in the unit test, but I guess it's good to have for the future)

---

*[drake/systems/sensors/rotary_encoders.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-KXHviyPGE9WB-IipsJt:-KXItqwH750432QDeHt8:bmd50yl) ([raw file](https://github.com/robotlocomotion/drake/blob/3fc2770b16b8b09e48078168323c14920748df38/drake/systems/sensors/rotary_encoders.h#L16)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary>
> interesting.  that is what I implemented at first, but i think the model i've used here is more correct... no?  (if you go negative, you don't get any ticks until you've crossed the first one). i've updated the doc.
</details>
i thought about it some more, and I guess the simpler solution (just using floor) is probably actually more correct.  i was worried about the case where you start up right at zero....  but I guess my version would have strange behavior (missing a tick) every time you cross zero.  I've changed back to the simpler form. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 11 files at r3, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/4247#-:-KXJV8Wwb8xevSdjYeyh:bnfp4nl)*
<!-- Sent from Reviewable.io -->
